R/tmb.R

```diff
@@ -357,12 +398,24 @@ h_mmrm_tmb_fit <- function(tmb_object,
   assert_data_frame(data)
   assert_class(formula_parts, "mmrm_tmb_formula_parts")
   assert_class(tmb_data, "mmrm_tmb_data")
-
+  
   tmb_report <- tmb_object$report(par = tmb_opt$par)
   x_matrix_cols <- colnames(tmb_data$x_matrix)
   visit_names <- levels(tmb_data$full_frame[[formula_parts$visit_var]])
-  cov <- tcrossprod(tmb_report$covariance_lower_chol)
-  dimnames(cov) <- list(visit_names, visit_names)
+  d <- dim(tmb_report$covariance_lower_chol)
+  cov <- lapply(
+    seq_len(d[2] / d[1]),
+    function(i) {
+      ret <- tcrossprod(tmb_report$covariance_lower_chol[, seq(1 + (i - 1) * d[1], i * d[1])])
+      dimnames(ret) <- list(visit_names, visit_names)
+      return(ret)
+    }
+  )
+  if (identical(tmb_data$n_groups, 1L)) {
+    cov <- cov[[1]]
+  } else {
+    names(cov) <- levels(tmb_data$subject_groups)
+  }
```

Comment:
this is quite complex. please factor out into helper function, helps with readability and testing